### PR TITLE
feat(pool): #556 Upsert Business Partners from Cleaning Result - refactor validation

### PR DESCRIPTION
**Pull request by rainer-exxcellent**
## Description

Upsert Business Partners from Cleaning Result - refactor validation 
prepare validation so that it can be called also by orchstrator

Solves: #556 

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [ ] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [ ] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

### CODE REVIEW

**Review by nicoprow — APPROVED** [review]
I like the addition of the validators in the request validation service. Looks good to me

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt [hunk 1]
```diff
@@ -272,117 +293,111 @@ class RequestValidationService(
...
-
-        return missingTypes.map {
-            ErrorInfo(
-                error,
-                "Legal Entity Identifier Type '$it' does not exist",
-                entityKey.getRequestKey()
-            )
+    inner class ValidateIdentifierLeTypesExists<ERROR : ErrorCode>(
+        requests: Collection<IBaseLegalEntityDto>,
+        private val error: ERROR) {
```
Comment (martinfkaeser) [inline bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:318]: Maybe call it `errorCode`?

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt [hunk 2]
```diff
@@ -117,51 +122,51 @@ class RequestValidationService(
...
+    ): Map<LegalEntityPartnerUpdateRequest, Collection<ErrorInfo<LegalEntityUpdateError>>> {
 
+        val legalEntityRequests = requests.map { it.legalEntity }
+        val duplicatesValidator = ValidateLegalEntityIdentifiersDuplicated(legalEntityRequests, LegalEntityUpdateError.LegalEntityDuplicateIdentifier)
         val existingLegalEntityBpns = legalEntityRepository.findDistinctByBpnIn(requests.map { it.bpnl }).map { it.bpn }.toSet()
+        val existingBpnValidator = ValidateUpdateBpnExists(existingLegalEntityBpns, LegalEntityUpdateError.LegalEntityNotFound)
+        val legalFormValidator = ValidateLegalFormExists(legalEntityRequests,LegalEntityUpdateError.LegalFormNotFound)
+        val identifierValidator = ValidateIdentifierLeTypesExists(legalEntityRequests, LegalEntityUpdateError.LegalEntityIdentifierNotFound)
 
         return requests.flatMap { request ->
```
Comment (martinfkaeser) [inline bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:141]: Can't you use
```suggestion
        return requests.associateWith { request -> ...
           validationErrors
        }
```
without `flatMap` and `groupBy`?

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt [hunk 3]
```diff
@@ -272,117 +293,111 @@ class RequestValidationService(
...
-        val requestedTypes = request.identifiers.map { it.type }
-        val missingTypes = requestedTypes - existingTypes
-
-        return missingTypes.map {
-            ErrorInfo(
-                error,
-                "Legal Entity Identifier Type '$it' does not exist",
-                entityKey.getRequestKey()
-            )
+    inner class ValidateIdentifierLeTypesExists<ERROR : ErrorCode>(
```
Comment (martinfkaeser) [inline bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:316]: Maybe call them `Validator...` not `Validate...`

**Review by martinfkaeser — CHANGES_REQUESTED** [review]
I like the Validators approach!
I added some suggestions for simplifing the code.
As you are already working on a follow-up PR, you best add your changes in the next PR.

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt [hunk 4]
```diff
@@ -484,33 +512,49 @@ class RequestValidationService(
...
     )
+
+    interface LegalEntityBridge {
+
+        val legalEntity: IBaseLegalEntityDto?
+        val request : RequestWithKey
+    }
+
+    companion object {
+        fun <TError : ErrorCode, TRequest : RequestWithKey> mergeErrorMaps(
```
[the remark this answers is not in the document]
Comment (rainer-exxcellent) [reply bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:538]: in a companion it is better to test and to reuse

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt (hunk 3 above)
Comment (rainer-exxcellent) [reply bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:316]: In this step the class is called like the refactored methods to make the refactoring clearer.
Renaming could be done in the next step

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt [hunk 5]
```diff
@@ -172,58 +174,87 @@ class RequestValidationService(
...
         val existingSiteBpns = siteRepository.findDistinctByBpnIn(requestedSiteBpns).map { it.bpn }.toSet()
+        val existingBpnValidator = ValidateUpdateBpnExists(existingSiteBpns, SiteUpdateError.SiteNotFound)
 
-        val addressDuplicateIdentifierCandidates = getAddressDuplicateIdentifierCandidates(mainAddressRequests)
+        return requests.flatMap { request ->
+            val validationErrors = existingBpnValidator.validate(request.bpns)
+
+            validationErrors.map { Pair(request, it) }
+
+        }.groupBy({ it.first }, { it.second })
```
Comment (martinfkaeser) [inline bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:239]: You could simplify the whole block to
```suggestion
        return requests.associateWith { request ->
            existingBpnValidator.validate(request.bpns)
        }
```

There are several cases in this class that could be replaced similarly where currently 
`.flatMap{ ... }.groupBy({ it.first }, { it.second })`
is used.

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (martinfkaeser) [reply bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:550]: Seems this `mergeErrorMaps()` function got duplicated accidentally?

By the way, you could also move this function out of the service, as a free function, as an alternative to a function in a companion object.

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt [hunk 6]
```diff
@@ -393,87 +410,102 @@ class RequestValidationService(
...
-            BusinessPartnerType.LEGAL_ENTITY -> validateParentBpnExists(
-                request.bpnParent,
-                request.index,
-                existingLegalEntities,
-                AddressCreateError.LegalEntityNotFound
-            )
+    private inner class ValidateAddressParent(requests: Collection<AddressPartnerCreateRequest>) {
+
+        val requestsWithParentType = requests.map { Pair(it, bpnIssuingService.translateToBusinessPartnerType(it.bpnParent)) }
+        val requestsByParentType = requestsWithParentType.groupBy({ it.second }, { it.first })
```
Comment (martinfkaeser) [inline bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:487]: You could simplify: 
```suggestion
      val requestsByParentType = requests.groupBy { bpnIssuingService.translateToBusinessPartnerType(it.bpnParent) }
```

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt (hunk 5 above)
Comment (rainer-exxcellent) [reply bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:239]: This is already done in the next implementation step. 
After the associateWith  a  `.filterValues { it.isNotEmpty() } ` is needed

File: bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt (hunk 6 above)
Comment (rainer-exxcellent) [reply bpdm-pool/src/main/kotlin/org/eclipse/tractusx/bpdm/pool/service/RequestValidationService.kt:487]: this is also already done in the next step

**Review by martinfkaeser — APPROVED** [review]

**Pull request merged**